### PR TITLE
[EHN] EEGMiner

### PR DESCRIPTION
**Pull request by bruAristimunha**
Small benchmark

Cross-Session
image
image

Cross-Subject:
image

### CODE REVIEW

File: braindecode/models/eegminer.py [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+    x_u_ii = torch.matmul(x_u[..., 1], x_u[..., 1].transpose(-2, -1))
+    x_u_ri = torch.matmul(x_u[..., 0], x_u[..., 1].transpose(-2, -1))
+    x_u_ir = torch.matmul(x_u[..., 1], x_u[..., 0].transpose(-2, -1))
+    r = x_u_rr + x_u_ii
+    i = x_u_ri - x_u_ir
+    time = amp.shape[-1]
+    plv = 1 / time * torch.sqrt(r**2 + i**2 + 1e-6)
+
+    return plv
+
```
Comment (bruAristimunha) [inline braindecode/models/eegminer.py:58]: I think I should be in functional/function

File: braindecode/models/eegminer.py [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+        method: str = "plv",
+        n_chans=None,
+        n_outputs=None,
+        n_times=None,
+        chs_info=None,
+        input_window_seconds=None,
+        sfreq=None,
+        # model related
+        filter_f_mean=[23.0, 23.0],
+        filter_bandwidth=[44.0, 44.0],
```
Comment (bruAristimunha) [inline braindecode/models/eegminer.py:275]: Tuple here

File: braindecode/models/eegminer.py [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+    x[..., 0, :] = x[..., 0, :] / 2.0  # Don't multiply the DC-term by 2
+    x = F.pad(
+        x, [0, 0, 0, x.shape[-2] - 2]
+    )  # Fill Fourier coefficients to retain shape
+    x = torch.view_as_complex(x)
+    x = torch.fft.ifft(x, norm=None, dim=-1)  # returns complex signal
+    x = torch.view_as_real(x)
+
+    return x
+
```
Comment (bruAristimunha) [inline braindecode/models/eegminer.py:40]: Test here

File: braindecode/models/eegminer.py (hunk 1 above)
Comment (bruAristimunha) [inline braindecode/models/eegminer.py:58]: This implementation is too verbose e overcomplicated, maybe rewrite with Rearrange layers or something else.

File: braindecode/models/eegminer.py [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+            )
+            self.n_features = self.n_filters * self.n_chans * (self.n_chans - 1) // 2
+            self.ensure_dim = nn.Identity()
+        elif self.method == "plv":
+            self.method_forward = partial(self._apply_plv, n_chans=self.n_chans)
+            self.ensure_dim = Rearrange("... d -> ... 1 d")
+            self.n_features = self.n_filters * self.n_chans * (self.n_chans - 1) // 2
+
+        # Classifier
+        self.ft_bn = nn.BatchNorm1d(self.n_features, affine=False)
```
Comment (bruAristimunha) [inline braindecode/models/eegminer.py:343]:
```suggestion
        self.batch_layer = nn.BatchNorm1d(self.n_features, affine=False)
```

File: braindecode/models/eegminer.py [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+        # Apply Gaussian filters in frequency domain
+        # x -> (batch, electrodes * filters, time)
+        x = self.filter(x)
+
+        x = self.method_forward(x=x, batch=batch)
+        # Classifier
+        # Note that the order of dimensions before flattening the feature vector is important
+        # for attributing feature weights during interpretation.
+        x = x.reshape(batch, self.n_features)
+        x = self.ft_bn(x)
```
Comment (bruAristimunha) [inline braindecode/models/eegminer.py:360]:
```suggestion
        x = self.batch_layer(x)
```

File: braindecode/models/eegminer.py [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+        x = x.abs()
+
+        # Get upper triu of symmetric connectivity matrix
+        triu = torch.triu_indices(n_chans, n_chans, 1)
+        x = x[:, triu[0], triu[1], :]
+
+        return x
+
+    @staticmethod
+    def _apply_plv(x, n_chans, batch=None):
```
Comment (bruAristimunha) [inline braindecode/models/eegminer.py:390]: We are not using batch here too

File: braindecode/models/eegminer.py [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+        n_chans=None,
+        n_outputs=None,
+        n_times=None,
+        chs_info=None,
+        input_window_seconds=None,
+        sfreq=None,
+        # model related
+        filter_f_mean=[23.0, 23.0],
+        filter_bandwidth=[44.0, 44.0],
+        filter_shape=[2.0, 2.0],
```
Comment (bruAristimunha) [inline braindecode/models/eegminer.py:276]:
```suggestion
        filter_f_mean=(23.0, 23.0),
        filter_bandwidth=(44.0, 44.0),
        filter_shape=(2.0, 2.0),
```

File: docs/whats_new.rst [hunk 8]
```diff
@@ -53,6 +53,7 @@ Enhancements
 - Adding :class:`braindecode.models.SyncNet` (:gh:`643` by `Bruno Aristimunha`_ )
 - Creating the `FilterBanklayer` module for new models (:gh:`656` by `Bruno Aristimunha`_ )
 - Including PytorchAudio as dependency and remove copied code (:gh:`661` by `Bruno Aristimunha`_)
+- Adding :class:`braindecode.models.EEGMiner` (:gh:`643` by `Bruno Aristimunha`_ )
```
Comment (bruAristimunha) [inline docs/whats_new.rst:56]:
```suggestion
- Adding :class:`braindecode.models.EEGMiner` (:gh:`667` by `Bruno Aristimunha`_ )
```

File: braindecode/models/eegminer.py [hunk 9]
```diff
@@ -0,0 +1,400 @@
...
+        # Note that the order of dimensions before flattening the feature vector is important
+        # for attributing feature weights during interpretation.
+        x = x.reshape(batch, self.n_features)
+        x = self.ft_bn(x)
+        x = self.final_layer(x)
+
+        return x
+
+    @staticmethod
+    def _apply_mag_forward(x, batch=None):
```
[the remark this answers is not in the document]
Comment (bruAristimunha) [reply braindecode/models/eegminer.py:196]: Do you know a better way to do this, @sylvchev?

File: braindecode/models/eegminer.py [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+
+        return x
+
+    @staticmethod
+    def _apply_plv(x, n_chans, batch=None):
+        # Compute PLV connectivity
+        # x -> (batch, electrodes, electrodes, filters)
+        x = x.transpose(-4, -3)  # swap electrodes and filters
+        x = F.plv_time(x, forward_fourier=False)
+        x = x.transpose(-3, -2).transpose(-2, -1)  # move filter channels to the end
```
Comment (sylvchev) [inline braindecode/models/eegminer.py:225]: Same here, is it possible to avoid the 2 transposes?

File: braindecode/models/eegminer.py [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+        x = x * x
+        x = x.mean(dim=-1)
+        x = torch.sqrt(x)
+        return x
+
+    @staticmethod
+    def _apply_corr_forward(x, batch, n_chans, n_filters, n_times):
+        x = x.reshape(batch, n_chans, n_filters, n_times).transpose(-3, -2)
+        x = (x - x.mean(dim=-1, keepdim=True)) / torch.sqrt(
+            x.var(dim=-1, keepdim=True) + 1e-6
```
Comment (sylvchev) [inline braindecode/models/eegminer.py:207]: Could you add the 1e-6 as an arg?

File: braindecode/models/eegminer.py [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+        x = x[:, triu[0], triu[1], :]
+
+        return x
+
+    @staticmethod
+    def _apply_plv(x, n_chans, batch=None):
+        # Compute PLV connectivity
+        # x -> (batch, electrodes, electrodes, filters)
+        x = x.transpose(-4, -3)  # swap electrodes and filters
+        x = F.plv_time(x, forward_fourier=False)
```
[the remark this answers is not in the document]
Comment (bruAristimunha) [reply braindecode/models/eegminer.py:224]: EEGMiner has three exclusive options: plv, corr, or mag as estimators over the filters, not sequential steps.

Okay for checking the swap and filter.

File: braindecode/models/eegminer.py [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+        return x
+
+    @staticmethod
+    def _apply_corr_forward(x, batch, n_chans, n_filters, n_times):
+        x = x.reshape(batch, n_chans, n_filters, n_times).transpose(-3, -2)
+        x = (x - x.mean(dim=-1, keepdim=True)) / torch.sqrt(
+            x.var(dim=-1, keepdim=True) + 1e-6
+        )
+        x = torch.matmul(x, x.transpose(-2, -1)) / x.shape[-1]
+        x = x.transpose(-3, -2).transpose(-2, -1)  # move filter channels to the end
```
[the remark this answers is not in the document]
Comment (bruAristimunha) [reply braindecode/models/eegminer.py]: done @sylvchev

File: braindecode/models/eegminer.py (hunk 10 above)
Comment (bruAristimunha) [reply braindecode/models/eegminer.py]: done @sylvchev

**Comment by sylvchev** [comment]
Regarding the licence, you could add a NOTICE file to indicate that EEGMiner is CC by-nc

**Comment by mccorsi** [comment]
Hi @sylvchev, do have contraints on the FC matrices you want to use (ie should they have to be SPD...)? I might have some ideas but I'd to be sure that it makes sense for the project! :)

**Comment by bruAristimunha** [comment]
No constraints on the FC @mccorsi and @sylvchev

**Review by sylvchev — APPROVED** [review]
Nice work. Thanks @bruAristimunha !

**Comment by mccorsi** [comment]
Here is a non-exhaustive list of FC metrics :

- Coherence-based metrics: cf the one already implemented in pyriemann & used in FUCONE
- more generally, mne-python increased the set of spectral FC metrics: Coherence, Coherency, Imaginary part of Coherency, Canonical Coherency (CaCoh), Maximised Imaginary part of Coherency (MIC), Multivariate Interaction Measure (MIM), Phase-Locking Value (PLV), Corrected Imaginary PLV (ciPLV), Pairwise Phase Consistency (PPC), Phase Lag Index (PLI), Unbiased estimator of squared PLI, Directed PLI (DPLI), Weighted PLI (WPLI), Debiased estimator of squared WPLI, State-space Granger Causality (GC), State-space GC on time-reversed signals
- bicoherence: to measure the coherence from two different bins of frequency

I am currently working on metrics that would capture non-linearities but it is still under development. I will let you know when it is ready (and implemented in python)!

**Pull request merged**